### PR TITLE
Support 'commisioning' advertising in CHIP

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
We need to support 'operational', 'commisioning' and 'commisionable' mDNS advertising.
 
 #### Summary of Changes
 This adds 'commisioning' advertising support in the minMDNS advertising spec.

This builds upon #4177 so changes are on the larger side.

### CODE REVIEW

**Review by arunbharadwaj — COMMENTED** [review]
I have suggested a couple of changes - since this is adding "commisioning" advertising, there were a few instances in the code which referred to "commisionable" mode instead. Fixed those.

Also I have a patch which adds the "commisionable" mode on top of this PR. Once this is merged, I will send out the PR for that.

File: examples/common/chip-app-server/Server.cpp [hunk 1]
```diff
@@ -159,22 +159,62 @@ void InitServer(AppDelegate * delegate)
...
+            ChipLogError(Discovery, "Fabric ID not known. Using a default");
+            fabricId = 5544332211;
+        }
+
+        err = Mdns::ServiceAdvertiser::Instance().Advertise(Mdns::OperationalAdvertisingParameters()
+                                                                .SetFabricId(fabricId)
                                                                 .SetNodeId(chip::kTestDeviceNodeId)
                                                                 .SetPort(CHIP_PORT)
-#if INET_CONFIG_ENABLE_IPV4
+                                                                .EnableIpV4(true));
```
Comment (bzbarsky-apple) [inline examples/common/chip-app-server/Server.cpp:176]: Why is `EnableIpV4` becoming unconditional?

**Review by rwalker-apple — APPROVED** [review]

File: examples/common/chip-app-server/Server.cpp (hunk 1 above)
Comment (andy31415) [reply examples/common/chip-app-server/Server.cpp:176]: Testing only really ... if you enable ipv4 we advertise that we have it.
We could make it optional or have additional config variables.

Also harsh reality for me: in my wifi setup, somehow ipv4 is a ton more reliable than ipv6, so until we make our ipv6 solid and testable, having dual v4 and v6 is probably easier on developers.

File: examples/minimal-mdns/advertiser.cpp [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+} gOptions;
+
+using namespace chip::ArgParser;
+
+constexpr uint16_t kOptionEnableIpV4      = '4';
+constexpr uint16_t kOptionAdvertisingMode = 'm';
+
+constexpr uint16_t kOptionCommisioningShordDiscriminator = 's';
+constexpr uint16_t kOptionCommisioningLongDiscriminaotr  = 'l';
+constexpr uint16_t kOptionCommisioningVendorId           = 0x100; // v is used by 'version'
```
[the remark this answers is not in the document]
Comment (andy31415) [reply examples/minimal-mdns/advertiser.cpp:62]: I considered it, however -v and -V look very similar, so I figured I would keep the full text in this case.

Ideally I would like an example to not have the '-version' at all, however changing that seems a lot more work that sidetracks the entire thing. Maybe a followup.

**Review by mspang — APPROVED** [review]

**Comment by mspang** [comment]
@saurabhst @hawk248 @BroderickCarlin @chrisdecenzo @jelderton

**Review by saurabhst — APPROVED** [review]

File: src/app/server/Server.cpp [hunk 3]
```diff
@@ -162,20 +162,62 @@ void InitServer(AppDelegate * delegate)
...
+            productId = Optional<uint16_t>::Value(value);
+        }
+
+        if (ConfigurationMgr().GetSetupDiscriminator(discriminator) != CHIP_NO_ERROR)
+        {
+            ChipLogError(Discovery, "Setup discriminator not known. Using a default.");
+            discriminator = 840;
+        }
+
+        err = Mdns::ServiceAdvertiser::Instance().Advertise(Mdns::CommisioningAdvertisingParameters()
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/app/server/Server.cpp:212]: The reason for this is mainly the 'goto exit' pattern: either I declare the advertiser on top (ugly) or I need to add braces to make the compiler happy and not give me a `expansion of macro SuccessOrExit crosses initialization ...`.

I will add variables and surround the code with one more layer of {} .

**Pull request merged**